outlook/indico_outlook/plugin.py

```diff
@@ -203,34 +297,94 @@ def event_registration_form_deleted(self, registration_form, **kwargs):
+
+        # Users that have marked the event as favorite too
+        for user in event.favorite_of:
+            if self._user_tracks_favorite_events(user):
+                users_to_update.add(user)
+
+        for category in reversed(event.category.chain_query.all()):
+            for user in category.favorite_of:
+                if self._user_tracks_favorite_categories(user) and event.can_access(user):
+                    users_to_update.add(user)
+            # Stop once we reach the visibility horizon of the event
+            if category is event.category.real_visibility_horizon:
+                break
+
+        for user in users_to_update:
             self.logger.info('Event data change: updating %s in %r', user, event)
             self._record_change(event, user, OutlookAction.update)
 
     def event_times_changed(self, sender, obj, **kwargs):
         event = obj
-        for user in get_participating_users(event):
-            self.logger.info('Event time change: updating %s in %r', user, event)
-            self._record_change(event, user, OutlookAction.update)
+        changes = kwargs['changes']
+        del kwargs['changes']
+
+        self.logger.info('Event time change: updating %r: %r', event, changes)
+        self.event_updated(event, changes, **kwargs)
+
+    def event_created(self, event, **kwargs):
+        self.logger.info('Event created: %r', event)
+        # Piggyback on the event_updated handler to avoid duplicating the logic
+        self.event_updated(event, {'title': event.title}, **kwargs)
 
     def event_deleted(self, event, **kwargs):
-        for user in get_participating_users(event):
+        users_to_update = set()
+        for user in get_registered_users(event):
+            if self._user_tracks_registered_events(user):
+                users_to_update.add(user)
+        for user in event.favorite_of:
+            if self._user_tracks_favorite_events(user):
+                users_to_update.add(user)
+        for category in reversed(event.category.chain_query.all()):
+            for user in category.favorite_of:
+                if self._user_tracks_favorite_categories(user) and event.can_access(user):
+                    users_to_update.add(user)
+            # Stop once we reach the visibility horizon of the event
+            if category is event.category.real_visibility_horizon:
+                break
+
+        for user in users_to_update:
             self.logger.info('Event deletion: removing %s in %r', user, event)
-            self._record_change(event, user, OutlookAction.remove)
+            self._record_change(event, user, OutlookAction.force_remove)
 
     def _record_change(self, event, user, action):
         if is_event_excluded(event):
             return
         if 'outlook_changes' not in g:
             g.outlook_changes = []
+
+        if action == OutlookAction.remove:
+            # Only remove an event if the user *really* shouldn't have it in their calendar
```

Comment:
Do we really need the new `force_remove` action? You could just add a new `force` kwarg to `_record_change`, and skip the checks here if it's set to True. No need to record this persistently.